presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -882,7 +882,7 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
             }
 
             // do not optimize non-deterministic functions
-            if (optimize && (!function.isDeterministic() || hasUnresolvedValue(argumentValues))) {
+            if (optimize && (!function.isDeterministic() || hasUnresolvedValue(argumentValues) || node.getName().toString().equals("grouping"))) {
```

Comment:
AFAIK, non-deterministic also means that the function will not be pushed down. This would be a nice side-effect to have, but isn't the best way to achieve this.